README.md
# Etcd-discovery

This is a golang package for managing services over the decentralized key-value store etcd

> http://github.com/coreos/etcd

To install it:

`go get github.com/Scalingo/etcd-discovery/service/v7`

## API

### Register a service

```go
/*
 * First argument is the name of the service
 * Then a struct containing the service informations
 * The registeration will stop when the context will be canceled.
 * It will return the service uuid and a channel which will send back any modifications made to the service by the other host of the same service. This is usefull for credential synchronisation.
 */
ctx, cancel := context.WithCancel(context.Background())
registration := service.Register(
  ctx,
  "my-service",
  &service.Host{
    Hostname: "public-domain.dev",
    Ports: service.Ports{
      "http":  "80",
      "https": "443",
    },
    Critical:       true,
    User:           gopassword.Generate(10),
    Password:       gopassword.Generate(10),
    Public: true,
    PrivateHostname: "node-1.internal.dev",
    PrivatePorts: service.ports{
      "http":  "8080",
      "https": "80443",
    },
  })
```

This will create two different etcd keys:

* `/services/name_of_service/you-service-uuid-node-1.internal.dev` containing:
```json
{
   "name": "public-domain.dev",
   "service_name": "my-service",
   "user": "user",
   "password": "secret",
   "public": true,
   "private_hostname": "node-1.internal.dev",
   "private_ports": {
      "http": "8080",
      "https": "80443"
   },
   "critcal": true,
   "uuid": "your-service-uuid-node-1.internal.dev",
   "ports":{
      "http":"80",
      "https":"443"
   }
}
```

* `/service_infos/name_of_service` containing:
```json
{
   "name": "my-service",
   "critical": true,
   "hostname": "public-domain.dev",
   "user": "user",
   "password": "password",
   "ports": {
      "http": "80",
      "https": "443"
   },
   "public": true
}
```

### Subscribe to new service

When a service is added from another host, if you want your application to
notice it and communicating with it, it is necessary to watch these
notifications.

```go
newHosts := service.SubscribeNew("name_of_service")
for host := range newHosts {
  fmt.Println(host.Name, "has registered")
}
```

### Watch down services

```go
deadHosts := service.SubscribeDown("name_of_service")
for hostname := range deadHosts {
  fmt.Println(hostname, "is dead, RIP")
}
```

# Generate the mocks

Generate the mocks with:

```shell
for interface in $(grep --extended-regexp --no-message --no-filename "type .* interface" ./service/* | cut -d " " -f 2)
do
  mockgen -destination service/servicemock/gomock_$(echo $interface | tr '[:upper:]' '[:lower:]').go -package servicemock github.com/Scalingo/etcd-discovery/service $interface
done
```


CHANGELOG
# To Be Released

## v7.0.2

* Use github.com/gofrs/uuid for UUID generation
* Transform into a go module
* Use go.etcd.io/etcd instead of github.com/coreos/etcd

## v7.0.1

* Fix panic when loosing registration

## v7.0.0

* Use context.Context to handle lifecycle of the registration

## v6.0.2

* Fix: Auto-fill wont work if the hostname exists
* Reset the AfterIndex if we've lost the etcd-watcher

## v6.0.1

* Try to auto-fill the private_hostname and private_ports if the service is'nt public

## v6.0

* Add the Registration struct to help Register provide a more user firendly API

## v5.0.1

* Use the host private hostname if the hostname is not set for the uuid generation

## v5.0

* Add the notion of Public URL
* Synchronize host password
* Remove the update subscriber
* Improve the Get() API

## v4.1

* Add `name` field in service_info struct

## v4.0

* Add notion of service info, to define for instance if a service is critical or not

## v3.3

* Ability to get TLS certs/key/ca from environment directly without being stored in files
  `ETCD_TLS_INMEMORY=true`

## v3.2

* String method on array of Host

## v3.1

* Fix TLS client from environment variables

## v3.0

* Compatibility API v3
* Use officail go client

## v2.0

* Change way to represent interfaces :
  As a service can bind different interfaces (not only http),
  it has now to specify them.

## v1.0

* Simple implementation of Register/SubscribeNew/SubscribeDown
* Test for these features


service/host.go
package service

import (
	"errors"
	"fmt"
	"strings"

	errgo "gopkg.in/errgo.v1"
)

// Ports is a representation of the ports exposed by a host or a service.
// The key is the protocol name and the value is the port used for this protocol.
// Typical usage is:
// 	Ports{
//		"http":"80",
//		"https": "443",
//  }
type Ports map[string]string

// Hosts will represent a slice of hosts
type Hosts []*Host

func (hs Hosts) String() string {
	names := []string{}
	for _, h := range hs {
		if len(h.PrivateHostname) != 0 {
			names = append(names, h.PrivateHostname)
		} else {
			names = append(names, h.Hostname)
		}
	}
	return strings.Join(names, ", ")
}

// Host store all the host informations. This is also used to store the host in the etcd services directory.
type Host struct {
	// Hostname  must be a publicly available IP/FQDN if the service is public or a private IP/FQDN id the service is private.
	Hostname string `json:"name"`
	// Name of the service which this host store. This will be overwritten by the Register Function
	Name string `json:"service_name"`
	// Ports is the ports accessible on a public network if the service is Public or the ports accessible on a private network if the service is private
	Ports Ports `json:"ports"`
	// User name used to authenticate to this service. The Register function can override this at any time if the service is public
	User string `json:"user,omitempty"`
	// Password used to authenticate to this service. The Register function can override this at any time if the service is public
	Password string `json:"password,omitempty"`
	// Public is set to true if the service is public
	Public bool `json:"public,omitempty"`
	// PrivateHostname is the private IP/FQDN used to communicate with this service on the private network
	// This will defaults to Hostname if empty
	PrivateHostname string `json:"private_hostname,omitempty"`
	// PrivatePorts is the private ports used to communicate with this service on the private network.
	// This will defaults to Ports if empty
	PrivatePorts Ports `json:"private_ports,omitempty"`
	// Critical will be set to true is the service is critical
	Critical bool `json:"critical,omitempty"`
	// UUID is the service UUID, this must have the following pattern: uuid-PrivateHostname
	UUID string `json:"uuid,omitempty"`
}

// URL will return a valid url to contact this service on the specific protocol provided by the scheme parameter
func (h *Host) URL(scheme, path string) (string, error) {
	var url string
	var port string
	var ok bool
	if port, ok = h.Ports[scheme]; !ok {
		return "", errors.New("unknown scheme")
	}

	if h.User != "" {
		url = fmt.Sprintf("%s://%s:%s@%s:%s%s",
			scheme, h.User, h.Password, h.Hostname, port, path,
		)
	} else {
		url = fmt.Sprintf("%s://%s:%s%s",
			scheme, h.Hostname, port, path,
		)
	}
	return url, nil
}

// PrivateURL will provide a valid url to contact this service on the Private network
// this method will fallback to the URL method if the host does not provide any PrivateURL
func (h *Host) PrivateURL(scheme, path string) (string, error) {
	if len(h.PrivateHostname) == 0 {
		url, err := h.URL(scheme, path)
		if err != nil {
			return "", errgo.Mask(err)
		} else {
			return url, nil
		}
	}
	var url, port string
	var ok bool
	if port, ok = h.PrivatePorts[scheme]; !ok {
		return "", errors.New("unknown scheme")
	}

	if len(h.User) != 0 {
		url = fmt.Sprintf("%s://%s:%s@%s:%s%s",
			scheme, h.User, h.Password, h.PrivateHostname, port, path,
		)
	} else {
		url = fmt.Sprintf("%s://%s:%s%s",
			scheme, h.PrivateHostname, port, path)
	}
	return url, nil
}

// HostResponse is the interface used to provide a single host response.
// This interface provide a standard API used fot method chaining like:
// 	Get("my-service").First().Url()
//
// To provide such API errores need to be stored and sent at the last moment.
// To do so each "final" methods (like URL or Host) will check if the Response is errored before
// continuing to their own logic.
type HostResponse interface {
	Err() error
	Host() (*Host, error)
	URL(scheme, path string) (string, error)
	PrivateURL(scheme, path string) (string, error)
}

// GetHostResponse is the HostResponse Implementation of the HostResponse interface used by the the Service methods.
// This only provide the error wrapping logic, all the actaul logic for these method are done by the Host struct.
type GetHostResponse struct {
	err  error
	host *Host
}

// Err will return an error if an error happened on any previous steps.
func (q *GetHostResponse) Err() error {
	return q.err
}

// Host will return the host represented by this error.
func (q *GetHostResponse) Host() (*Host, error) {
	if q.err != nil {
		return nil, errgo.Mask(q.err)
	}

	return q.host, nil
}

// URL will return the public URL of this host
func (q *GetHostResponse) URL(scheme, path string) (string, error) {
	if q.err != nil {
		return "", errgo.Mask(q.err)
	}
	url, err := q.host.URL(scheme, path)
	if err != nil {
		return "", errgo.Mask(err)
	}
	return url, nil
}

// PrivateURL will return the private URL of this host
func (q *GetHostResponse) PrivateURL(scheme, path string) (string, error) {
	if q.err != nil {
		return "", errgo.Mask(q.err)
	}
	url, err := q.host.PrivateURL(scheme, path)
	if err != nil {
		return "", errgo.Mask(err)
	}
	return url, nil
}
